File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/constant/StringConstant.java
package xyz.wagyourtail.commons.parsers.impl.constant;

import lombok.val;
import xyz.wagyourtail.commons.core.StringUtils;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;

public class StringConstant extends StringData.OnlyRaw<Data.SingleContent<String>> {


    private StringConstant(String content) {
        super(content, StringConstant::getContentChecked);
    }

    public StringConstant(CharReader<?> content) {
        super(content, StringConstant::getContentChecked);
    }

    public static StringConstant parse(String rawContent) {
        StringCharReader reader = new StringCharReader(rawContent);
        val number = new StringConstant(reader);
        reader.expectEOS();
        return number;
    }

    public static StringConstant unchecked(String rawContent) {
        return new StringConstant(rawContent);
    }

    public String getValue() {
        val content = getRawContent();
        return StringUtils.translateEscapes(content.substring(1, content.length() - 1));
    }

    private static SingleContent<String> getContentChecked(CharReader<?> reader) {
        return new SingleContent<>(reader.takeString(CharReader.TAKE_STRING_NO_TRANSLATE_ESCAPES, "\""));
    }

}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/constant/StringConstant.kt
package xyz.wagyourtail.commonskt.parsers.impl.constant

import xyz.wagyourtail.commonskt.parsers.Data
import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.reader.CharReader
import xyz.wagyourtail.commonskt.utils.translateEscapes

class StringConstant(rawContent: String) : StringData.OnlyRaw<Data.SingleContent<String>>(rawContent) {

    constructor(content: SingleContent<String>) : this(content.toString())

    companion object : StringDataBuilder<StringConstant, Data.SingleContent<String>> {

        override fun invoke(rawContent: CharReader<*>): StringConstant {
            return StringConstant(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
            return SingleContent(reader.takeString(noTranslateEscapes = true, quote = "\""))
        }

    }

    val value: String
        get() = rawContent.substring(1, rawContent.length - 1).translateEscapes()

    override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
        return Companion.checkedBuildContent(reader)
    }


}